freqtrade/optimize/backtesting.py

```diff
@@ -325,19 +325,29 @@ def backtest(self, args: Dict) -> DataFrame:
             pairs.append(pair)
 
         lock_pair_until: Dict = {}
+        indexes: Dict = {}
         tmp = start_date + timedelta(minutes=self.ticker_interval_mins)
-        index = 0
+
         # Loop timerange and test per pair
         while tmp < end_date:
             # print(f"time: {tmp}")
+
             for i, pair in enumerate(ticker):
+                if pair not in indexes:
+                    indexes[pair] = 0
+
                 try:
-                    row = ticker[pair][index]
+                    row = ticker[pair][indexes[pair]]
                 except IndexError:
                     # missing Data for one pair ...
                     # Warnings for this are shown by `validate_backtest_data`
                     continue
 
+                if row.date > Timestamp(tmp.datetime):
+                    continue
```

Comment:
2 questions about these 2 lines:
a) why cast to timestamp - python can compare datetime objects natively
b) I'm not sure I understand the reason for this comparison / skip statement ... is this to continue looping until your data actually starts ?

Reply:
a) oh sorry I used it in some tests and I forgot to remove it.
b) yes, this is basically the main change. We should evaluate a pair only if it's start date is >= `tmp`, otherwise concurrent trades can happen (with `max_open_trades=1`).

Reply:
so in case you have missing data at the start for one pair, that pair would have a final index of 900 - while all others have 1100 (assuming 200 candles at the start are missing) ? 

I like the idea :+1:

Reply:
exactly. In this way a scenario like the one I described in the issue does not occur; pairs B and C are not bought since the first pair (in chronological order) is A. This also reflects dry run/live behaviour.

Reply:
I have difficulties understanding this loop to be honest. I feel you guys are much into it but as someone new, it is really difficult to understand it. there is almost no comment explaining the logic and the reason behind it. and so many if/else without knowing the goal of it beforehand. It would be great if you could kindly:

- comment the purpose of each variable before declaring it.
- explain the logic.
- encapsulate the logic somehow. maybe not in a new function but with the help of new lines before and after the logic. The following is just a proposition:
```
# beginning of detecting missing data
...
# end of detecting missing data
```
- Do the same for the tests, `multi`does not mean much to a new comer.